File: js/room.js
class Room {
    constructor(room_id) {
        this.room_id = room_id;
        this.raw     = rooms[this.room_id];
        this.name    = this.raw.name;
        this.owner   = this.raw.owner;
        this.tiles   = [];
        this.props   = [];
        this.warps   = [];
        this.signs   = [];
        this.north   = this.raw.north;
        this.east    = this.raw.east;
        this.south   = this.raw.south;
        this.west    = this.raw.west;
        this.color   = this.raw.color;
        this.weather = this.raw.weather;
        this.height  = TILESIZE * MAP_HEIGHT;
        this.width   = TILESIZE * MAP_WIDTH;
        // Set warps
        for (var k = 0; k < this.raw.warps.length; k++) {
            var warp = this.raw.warps[k];
            var warp_x  = warp[0];
            var warp_y  = warp[1];
            var room_id = warp[2];
            var room_x  = warp[3];
            var room_y  = warp[4];
            if (room_id != null) {
                this.warps.push(
                    new Warp(warp_x, warp_y, room_id, room_x, room_y)
                );
            }
        }
        // Set props
        for (var k = 0; k < this.raw.props.length; k++) {
            var prop   = this.raw.props[k];
            var prop_x = prop[0];
            var prop_y = prop[1];
            var text   = prop[2];
            if (text != "") {
                this.props.push(
                    new Prop(prop_x, prop_y, text)
                );
            }
        }
        // Set signs
        for (var k = 0; k < this.raw.signs.length; k++) {
            var sign   = this.raw.signs[k];
            var sign_x = sign[0];
            var sign_y = sign[1];
            var text   = sign[2];
            if (text != "") {
                this.signs.push(
                    new Sign(sign_x, sign_y, text)
                );
            }
        }
        // Set tiles
        for (var j = 0; j < MAP_HEIGHT; j++) {
            for (var i = 0; i < MAP_WIDTH; i++) {
                var index = (j * MAP_WIDTH) + i;
                var type  = this.raw.tiles[index];
                this.tiles.push(
                    new Tile(i, j, type)
                );
            }
        }
    }

    removeProp(i, j) {
        var index = -1;
        for (var k = 0; k < this.raw.props.length; k++) {
            var prop = this.raw.props[k];
            if ((i == prop[0]) && (j == prop[1])) {
                index = k;
            }
        }
        if (index > -1) {
            this.raw.props.splice(index, 1);
        }
    }
    
    setProp(i, j, type) {
        var index = -1;
        for (var k = 0; k < this.raw.props.length; k++) {
            var prop = this.raw.props[k];
            if ((i == prop[0]) && (j == prop[1])) {
                index = k;
            }
        }
        if (index == -1) {
            this.raw.props.push([i, j, type]);
        } else {
            this.raw.props[index] = [i, j, type];
        }
        this.update();
    }

    update() {
        var data = {
            room_id: this.room_id,
            name: this.name,
            owner: this.owner,
            color: this.color,
            tiles: this.raw.tiles,
            props: this.raw.props,
            signs: this.raw.signs,
            warps: this.raw.warps,
            weather: this.weather
        };
        var ref = database.ref('mmo/rooms');
        ref.child(this.room_id).set(data, this.gotData);
    }

    display() {
        background(this.color);
        //document.getElementById("game").style.backgroundColor = this.color;
        document.getElementById("room_name").innerHTML = this.name;
        for (var k = 0; k < this.tiles.length; k++) {
            this.tiles[k].display();
        }
        for (var k = 0; k < this.signs.length; k++) {
            this.signs[k].display();
        }
        for (var k = 0; k < this.props.length; k++) {
            this.props[k].display();
        }
    }

    tick() {
        for (var index in this.props) {
            this.props[index].tick();
        }
    }

    interact(i, j) {
        for (var index in this.props) {
            var prop = this.props[index];
            if ((prop.i == i) && (prop.j == j)) {
                prop.interact();
            }
        }
    }
}

function getRoomData(room_id) {
    if (room_id in rooms) {
        return rooms[room_id].tiles;
    } else {
        return [];
    }
}

File: rb/room.rb
KEY_PATH = "./files/key.json"

REF = [
  [' ', 'GROUND'],
  ['G', 'GROUND'],
  ['"', 'GRASS'],
  ['*', 'FLOWERS'],
  ['O', 'OAK_TREE'],
  ['F', 'FIR_TREE'],
  ['P', 'PINE_TREE'],
  ['C', 'CHERUB_TREE'],
  ['─', 'WALL_EW'],
  ['┄', 'WALL_EW_ALT'],
  ['|', 'WALL_NS'],
  ['┌', 'WALL_ES'],
  ['┐', 'WALL_SW'],
  ['└', 'WALL_NE'],
  ['┘', 'WALL_NW'],
  ['┬', 'WALL_ESW'],
  ['┴', 'WALL_NEW'],
  ['╵', 'WALL_NS_N'],
  ['╷', 'WALL_NS_S'],
  ['^', 'STAIRS_UP'],
  ['˅', 'STAIRS_DOWN'],
  ['S', 'SIGN'],
  ['X', 'WARP_BLUE'],
  ['F', 'FIRE']
]

class Room
  attr_accessor :id, :data, :w, :h, :color, :weather
  attr_accessor :north, :east, :south, :west

  def initialize(room_id, map = nil)
    @id, @color = room_id, "#222323"
    @w, @h, @north, @east, @south, @west = 24, 16, nil, nil, nil, nil
    @data, @warps, @signs = Array.new(@w * @h), [[-1, -1, -1, -1, -1]], [[-1, -1, ""]]
    @weather = true
    @data = map.split("") unless (map == nil) || (map.length != @w * @h)
  end

  def ix(i, j); j * @w + i; end

  def fill(t); @data.map! { t }; end

  def tile(i, j, t); @data[ix(i, j)] = t; end
  
  def warp(i, j, r, x, y); @warps << [i, j, r, x, y]; end
  
  def sign(i, j, s); @signs << [i, j, s]; end

  def col(i, t); (0...@h).each { |j| @data[ix(i, j)] = t }; end

  def row(j, t); (0...@w).each { |i| @data[ix(i, j)] = t }; end
  
  def random_fill(t, r); @data.map! { |k| rand < r ? t : k }; end

  def scatter_fill(t); @data.map! { |k| t[(rand * t.length).floor] }; end

  def fill_area(a, t); (a[1]..a[3]).each { |j| (a[0]..a[2]).each { |i| @data[ix(i, j)] = t } }; end

  def scatter_area(a, t); (a[1]..a[3]).each { |j| (a[0]..a[2]).each { |i| @data[ix(i, j)] = t[(rand * t.length).floor] } }; end

  def overlay(map); map.split("").each_with_index { |k, i| @data[i] = k unless k == " " }; end

  def reformat(); REF.each { |a, b| @data.map! { |c| (a == c) ? b : c } }; end

  def draw(); (0...@h).each { |j| puts (0...@w).map { |i| @data[ix(i, j)][0] + " " }.join }; end

  def export()
    {
      :room_id => @id,
      :color => @color,
      :data => @data,
      :warps => @warps,
      :signs => @signs,
      :north => @north,
      :east => @east,
      :south => @south,
      :west => @west,
      :weather => @weather
    }
  end

  def save()
    base_uri = 'https://machin-dev.firebaseio.com'
    auth_token = File.open(KEY_PATH).read
    firebase = Firebase::Client.new(base_uri, auth_token)
    #firebase.set(path, id => data)
    response = firebase.set("mmo/rooms/#{@id}", export())
    puts response.success? ? "Room \"#{@id}\" added to database" : "Room could not be added"
    return response.success?
  end
end

def get_rooms()
  base_uri = 'https://machin-dev.firebaseio.com'
  auth_token = File.open(KEY_PATH).read
  firebase = Firebase::Client.new(base_uri, auth_token)
  result = []
  return firebase.get("mmo/rooms").body
end